src/Analyzer.Utilities/JTokenExtensions.cs

```diff
@@ -13,107 +15,320 @@ namespace Microsoft.Azure.Templates.Analyzer.Utilities
-                        {
-                            token = childProperty.Value;
-                            foundChild = true;
-                            break;
-                        }
                     }
                 }
+
                 if (!foundChild)
                 {
                     switch (behaviorWhenNotFound)
                     {
                         case InsensitivePathNotFoundBehavior.Error:
-                            throw new Exception($"JSON path was unresolved at {string.Join(",", pathSoFar)}", lastException);
+                            throw new Exception($"JSON path was unresolved at {string.Join(".", tokenContext.ResolvedPath)}", lastException);
                         case InsensitivePathNotFoundBehavior.LastValid:
-                            return token;
-                        case InsensitivePathNotFoundBehavior.Null:
-                            return null;
+                            yield return tokenContext.Token;
+                            break;
+
+                        // Paths not fully resolved are ignored for InsensitivePathNotFoundBehavior.Null.
+                        // If the path doesn't contain wildcards, a single null may be returned below.
+                    };
+                }
+            }
+
+            // Return null if full, non-wildcard path is not found
+            // and behaviorWhenNotFound is Null
+            if (behaviorWhenNotFound == InsensitivePathNotFoundBehavior.Null
+                && !(pathContainsWildcard || fullPathFound))
+            {
+                yield return null;
+            }
+        }
+
+        /// <summary>
+        /// Helper method for InsensitiveTokens().
+        /// Adds all child properties of the token context to the queue to be processed later.
+        /// For any child property which has a value of an array, all elements of the array
+        /// are added to the queue.
+        /// </summary>
+        private static void AddAllChildrenToQueue(Queue<InsensitiveTokenContext> tokenQueue, InsensitiveTokenContext tokenContext)
+        {
+            foreach (var child in tokenContext.Token.Children())
+            {
+                if (child is JProperty childProperty)
+                {
+                    if (childProperty.Value is JArray jArray)
+                    {
+                        AddAllArrayElementsToQueue(tokenQueue, jArray, tokenContext);
+                        continue;
                     }
+
+                    EnqueueChildJToken(tokenQueue, childProperty.Value, tokenContext);
                 }
             }
+        }
+
+        /// <summary>
+        /// Helper for InsensitiveTokens().
+        /// Adds all elements of a JArray to the queue to be processed later.
+        /// </summary>
+        private static void AddAllArrayElementsToQueue(Queue<InsensitiveTokenContext> tokenQueue, JArray jArray, InsensitiveTokenContext parentContext)
```

Comment:
Do you want to call this `AddAllJArrayElementsToQueue` instead?

Reply:
I can do that.